frontend/src/components/common/Terminal.tsx

```diff
@@ -40,43 +54,67 @@ export default function Terminal(props: TerminalProps) {
   const [terminalContainerRef, setTerminalContainerRef] = React.useState<HTMLElement | null>(null);
   const [container, setContainer] = React.useState<string | null>(null);
   const { t } = useTranslation('resource');
+  const execRef = React.useRef<any>(null);
+  const fitAddonRef = React.useRef<any>(null);
+  const xtermRef = React.useRef<any>(null);
+  const [shells, setShells] = React.useState({
+    available: getAvailableShells(),
+    currentIdx: 0,
+    automaticallySwitch: true,
+  });
+
+  function getAvailableShells() {
+    const selector = item.spec?.nodeSelector || {};
+    const os = selector['kubernetes.io/os'] || selector['beta.kubernetes.io/os'];
+    if (os === 'linux') {
+      return ['bash', 'sh'];
+    } else if (os === 'windows') {
+      return ['powershell.exe', 'cmd.exe'];
+    }
+    return ['powershell.exe', 'bash', 'cmd.exe', 'sh'];
+  }
 
   function getDefaultContainer() {
     return item.spec.containers.length > 0 ? item.spec.containers[0].name : '';
   }
 
   // @todo: Give the real exec type when we have it.
-  function setupTerminal(
-    containerRef: HTMLElement,
-    xterm: XTerminal,
-    fitAddon: FitAddon,
-    exec: any
-  ) {
+  function setupTerminal(containerRef: HTMLElement, xterm: XTerminal, fitAddon: FitAddon) {
     if (!containerRef) {
       return;
     }
 
     xterm.open(containerRef);
-    fitAddon.fit();
 
-    xterm.onKey(event => {
-      // For some reason, pressing a key can give us a "keydown" or "keypress" event...
-      if (!['keydown', 'keypress'].includes(event.domEvent.type)) {
-        return;
-      }
+    xterm.onData(data => {
+      send(0, data);
+    });
 
-      // Send a newline when pressing enter
-      const code = event.domEvent.key === 'Enter' ? '\n' : event.key;
+    xterm.onResize(size => {
+      console.log('xterm resized: ' + size);
+      send(4, `{"Width":${size.cols},"Height":${size.rows}}`);
+    });
 
-      // We just send the key strokes to the socket; the actual writing into the
-      // terminal will be done when the data arrives.
-      send(code, exec);
+    // Allow copy/paste in terminal
+    xterm.attachCustomKeyEventHandler(arg => {
+      if (arg.ctrlKey && arg.code === 'KeyC' && arg.type === 'keydown') {
+        const selection = xterm.getSelection();
+        if (selection) {
+          return false;
+        }
```

Comment:
This means Ctrl+C without a selection will still abort the process, right? If so, this is very good! (I don't want to break the regular Unix Ctrl+C shortcut)